### PR TITLE
Postgres raster temporal API support

### PR DESCRIPTION
**Pull request by elpaso**
## Temporal capabilities integration for Postgres rasters

This PR also contains a few bug fixes and speed optimizations, I'm planning to add a some more test cases before the final merge.

qgis-pg-raster-temporal

qgis-pgraster-temporal-default

Funded by: **ARPA Piemonte**

### CODE REVIEW

**Comment by elpaso** [comment]
@Samweli see my last commits, I cannot see any error here after my last commits, but you probably know how to reproduce the issues better than me.

**Comment by Samweli** [comment]
> @Samweli see my last commits, I cannot see any error here after my last commits, but you probably know how to reproduce the issues better than me.

@elpaso I have tested, I run into different issues, one is layer datasource is not available after static temporal range update.
Error_wms_temporal_update

File: python/core/auto_generated/qgsdataprovider.sip.in [hunk 1]
```diff
@@ -116,7 +116,7 @@ Gets the data source specification.
 .. versionadded:: 3.0
 %End
 
-    virtual QgsDataProviderTemporalCapabilities *temporalCapabilities();
+    virtual QgsDataProviderTemporalCapabilities *temporalCapabilities() const;
```
Comment (nyalldawson) [inline python/core/auto_generated/qgsdataprovider.sip.in:119]: There's discussion about this here: https://github.com/qgis/QGIS/pull/35466#discussion_r401269274

File: src/providers/postgres/raster/qgspostgresrasterprovider.cpp [hunk 2]
```diff
@@ -1185,6 +1228,46 @@ bool QgsPostgresRasterProvider::init()
     return false;
   }
 
+  // Temporal capabilities
+  // Setup temporal properties for layer
+  if ( mUri.hasParam( QStringLiteral( "temporalField" ) ) )
+  {
+    Q_ASSERT_X( temporalCapabilities(), "QgsPostgresRasterProvider::QgsPostgresRasterProvider", "Data provider temporal capabilities object does not exist" );
```
Comment (nyalldawson) [inline src/providers/postgres/raster/qgspostgresrasterprovider.cpp:1235]: I don't think this is explicitly required -- the temporalCapabilities object is always created by QgsRasterDataProvider

**Review by nyalldawson — COMMENTED** [review]
Looking good to me! 3.14 will be when QGIS enters the 4th dimension :smile: 

One thing you may want to consider -- the wms provider exposes the option of what to do when an exact time range match isn't possible. E.g. picking the closest possible value, using the time from the start of the range, the end of the range, etc. I think those options could also be useful here.

**Comment by elpaso** [comment]
@Samweli I cannot reproduce, can you please make sure you are testing exactly this branch? If you still have issues can you upload a project and the exact steps to reproduce the issue?

**Comment by Samweli** [comment]
> @Samweli I cannot reproduce, can you please make sure you are testing exactly this branch? If you still have issues can you upload a project and the exact steps to reproduce the issue?

Yes I'm using this branch, I have attached below screenshot of steps to reproduce invalid projection and disabled reference time option issues for a layer which have reference time in capabilities.
I have used https://geo.weather.gc.ca/geomet?layer=GDPS.ETA_TT layer.

wms_temporal_range_update_error

**Comment by timlinux** [comment]
So exciting to see this @elpaso ! Do you have a sample dataset / table dump that we can play with?

**Comment by elpaso** [comment]
@timlinux there isfew sample datasets in the testdata folder here: 

https://raw.githubusercontent.com/qgis/QGIS/0351cb203d58ca2492bfd99c693a96453bd223d0/tests/testdata/provider/postgresraster/raster_4326_time.sql

But I'm going to replace it soon with a smaller dataset for the CI tests.

**Comment by Samweli** [comment]
> @elpaso I have tested, I run into different issues, one is layer datasource is not available after static temporal range update.
> Error_wms_temporal_update

Hi @elpaso, I'm still getting the above issue.

**Pull request merged**